### ISSUE TITLE
Inconsistency in placement of comments as returned by OpDefOrDeclNode.getComment

### ISSUE THREAD

**Issue by konnov** [tlaplus/tlaplus#578]
We are now using `OpDefOrDeclNode.getComment` to parse annotations in TLA+ declarations, see [this PR](https://github.com/informalsystems/apalache/pull/503). While it works well, we have noticed a few inconsistencies in the placement of comments. For instance, in the recursive definitions, a comment has to be placed between `RECURSIVE ...` and the definition, whereas in the local definitions, a comment has to be placed in front of `LOCAL`.

Below you can find an example of how annotations should be placed in front of declarations. It would be great, if SANY was using a more consistent scheme for `getComment`.

```tla
-------- MODULE Annotations ----------
EXTENDS Integers

CONSTANT
  \* @type: Int;
  N

VARIABLE
  \* @type: Set(Int);
  set

\* @pure
\* @type: Int => Int;
Inc(n) == n + 1

\* @type: Int => Int;
LOCAL LocalInc(x) == x + 1

A(n) ==
  LET \* @pure
      \* @type: Int => Int;
      Dec(x) == x + 1
  IN
  Dec(n)

RECURSIVE Fact(_)
\* @tailrec
\* @type: Int => Int;
Fact(n) ==
  IF n <= 1 THEN 1 ELSE n * Fact(n - 1)

\* @tailrec
\* @type: Int -> Int;
FactFun[n \in Int] ==
  IF n <= 1 THEN 1 ELSE n * FactFun[n - 1]

======================================
```

**Comment by lemmy** [comment]
@konnov Are you going to provide a patch?

**Comment by konnov** [comment]
I would love to, but I think we first need a consensus on what style is seen consistent. Probably the easiest thing to do would be to place annotations after `LOCAL`.

**Comment by lemmy** [comment]
FWIW: Annotations appear to be an interesting alternative to the config file. Especially, when combined with multi-module .tla files:

```tla
---- MODULE spec ----
\* @Const(42)
CONSTANT N

VARIABLE v

Init == v = N

Next == v = v'

\* @BehaviorSpec
Spec == Init /\ [][Next]_v

\* @Invariant
Inv == v = 42

====

---- MODULE model ----

\* @StateContraint
StateConstraint == v = 42
====
```

**Comment by konnov** [comment]
Yes. That is a nice application of annoations!

**Comment by lemmy** [comment]
This suggests that we should think about namespaces for annotations. For example, `StateConstraint` should best be `tlc.StateConstraint` to indicate it's specific to TLC.

**Comment by konnov** [comment]
I would avoid dots in the annotation names. But yes, having a convention would not hurt.

**Comment by lemmy** [comment]
I'm not married to a particular format.  I just went with how namespaces work in Java.  What's your preference?

**Comment by konnov** [comment]
I would prefer just to have normal identifiers. It is true that modern annotations in Java are put in packages. As far as I remember, the old annotations (in comments) were just identifiers. I don't think that we need the complexity of modern Java annotations.

**Comment by lemmy** [comment]
"apalache.type" could be a normal identifier that would tell tools and users alike to what it applies.

**Comment by konnov** [comment]
Not according to JLS: https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8

**Comment by lemmy** [comment]
Why do we have to care about the JLS?  Anyway, make it "@apalache_type".

**Comment by lemmy** [comment]
@konnov Is this still needed?

**Comment by konnov** [comment]
Yes! We have not received complaints about `LOCAL` so far. Most likely, because nobody needed an annotation of a local operator so far.

However, several people were confused about the comments not falling through the `VARIABLE` and `CONSTANT` keywords. For instance, we require the users to write annotations like this:

```tla
CONSTANT
  \* @type: Int;
  N

VARIABLE
  \* @type: Set(Int);
  set
```

But some people want them like this:

```tla
\* @type: Int;
CONSTANT N

\* @type: Set(Int);
VARIABLE set
```

I don't know how hard it is to make the comment on top of `CONSTANT` or `VARIABLE` get propagated to the first name in the list. So we can live without this feature, but it would be nice to have.

**Comment by lemmy** [comment]
@ahelwer What are the rules WRT comments in https://github.com/tlaplus-community/tree-sitter-tlaplus?  Looking at the parse tree of, e.g. https://github.com/tlaplus/Examples/blob/master/specifications/DieHard/DieHard.tla, indicates that comments are more or less placed arbitrarily.

**Comment by ahelwer** [comment]
Yeah tree-sitter doesn't provide much control over where comments are actually placed in the parse tree.

**Comment by Calvin-L** [comment]
[](url)

**Comment by lemmy** [comment]
@Calvin-L Did you accidentally close this issue?

**Comment by Calvin-L** [comment]
Yes, it appears I clicked on the wrong button while browsing on mobile. Embarrassing!
